File: Project Euler/Problem/019/solution.c
// https://euler.synap.co.kr/problem=19
// https://projecteuler.net/problem=19

#include <stdio.h>
 
int main()
{
    int day = 1; 
    int year, month;
    int count = 0;
    int is_leap;
 
    int arr[2][12] = { 
        { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 },
        { 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 }
    };
 
    for (month = 1; month <= 12; ++month) 
    {
        day += arr[0][month - 1];
    }
    day = 1 + (7 - (day % 7));
 
    for (year = 1901; year <= 2000; ++year) 
    {
        is_leap = ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0);
 
        for (month = 1; month <= 12; ++month)
        {
            count += (day == 1);
 
            while (day <= arr[is_leap][month - 1])
            {
                day += 7;
            }
 
            day -= arr[is_leap][month - 1];
        }
    }
 
    printf("%d\n", count);
 
    return 0;
}

File: Project Euler/Problem/019/Solution.java
// https://euler.synap.co.kr/problem=19
// https://projecteuler.net/problem=19

public class Solution {
    
    public static void main(String[] args) {
        System.out.println(run());
    }
    
    public static String run() {
        int sunday;
        int count = 0;
        int today = 1;
        
        for(int month = 1; month <= 12; month++) {
            switch(month) {
                case 1: case 3:  case 5: case 7: case 8: case 10: case 12:
                    today += 31;
                    break;
                case 2:
                    today += 28;
                    break;    
                case 4: case 6: case 9: case 11:
                    today += 30;
                    break;
                    
            }
        }
        
        sunday = 1 + (7 - (today % 7));
        
        for(int year = 1901; year <= 2000; year++) {
            for(int month = 1; month <= 12; month++) {                
                if(sunday == 1) { count++; }
                
                switch(month) {
                    case 1: case 3: case 5: case 7: case 8: case 10: case 12:
                        while(sunday <= 31) { 
                            sunday += 7;
                        }
                        
                        sunday -= 31;
                        break;
                    case 2:
                        if(((year%4 == 0) && (year%100 != 0)) || (year%400 == 0)) {
                            while(sunday < 29) {
                                sunday += 7;
                            }
                            
                            sunday -= 29;
                        } else {
                            while(sunday < 28) {
                                sunday += 7;
                            }
                            
                            sunday -= 28;
                        }
                        break;
                    case 4: case 6: case 9: case 11:                    
                        while(sunday <= 30) {
                            sunday += 7;
                        }
                        
                        sunday -= 30;
                        break;
                }
            }
        }
        
        return Integer.toString(count);
    }
}
